Chapter-4-applying-tactical-domain-driven-design/README.adoc

```diff
@@ -40,29 +77,238 @@ IMPORTANT: It makes no sense to consider the Domain Model in typical CRUD module
+
+==== Main characteristics of entities:
+
+- They have a unique identity
+- They represent a business concept
+- They have behavior (methods)
+- They have state (properties)
+- They can be changed over time
+- They encapsulate business logic
+- They can raise domain events after creating or state change
+- They can be internal part of an aggregate root
+- They can be becoming aggregate root when needed of protecting invariants
+- In one bounded context concept can be modeled as a value object and in another as an entity
+
+[source,csharp]
+----
+public sealed class Annex : Entity
+{
+    public AnnexId Id { get; init; } // Unique Entity Id
+    public BindingContractId BindingContractId { get; init; }
+    public DateTimeOffset ValidFrom { get; init; } // State
+
+    // EF needs this constructor to create non-primitive types
+    private Annex() { }
+
+    private Annex(BindingContractId bindingContractId, DateTimeOffset validFrom)
+    {
+        Id = AnnexId.Create();
+        BindingContractId = bindingContractId;
+        ValidFrom = validFrom;
+
+        var @event = AnnexAttachedToBindingContractEvent.Raise(Id, BindingContractId, ValidFrom); // Raise domain event
+        RecordEvent(@event); 
+    }
+
+    internal static Annex Attach(BindingContractId bindingContractId, DateTimeOffset validFrom) =>
+        new(bindingContractId, validFrom); // Behavior method
+}
+----
+
+NOTE: You probably have heard about Anemic Domain Model. This is known as anti-pattern. It is an entity that has only properties and no behavior. It is acceptable when you have simple CRUD operations. In complex process, we recommend to encapsulate behavior (methods) in the domain entity insted having this logic in the service layer.
+
+[[value-object]]
+==== Value Object
+
+A **Value Object** represents a business concept without a lifecycle. Unlike entities, value objects lack identity. They are immutable, serving as explicit types that describe specific aspects of the domain. 
+
+They enhance our domain model's expressiveness, prevent invalid object state and helps us avoid primitive obsession.
+
+"Primitive Obsession" is a code smell where we use primitive types to represent domain concepts. 
+
+image::Assets/value_object_sets.png[300,300,align=center]
+
+For e.g you can treat _SSN_ as _string_. _String_ allow put to field every character, but _SSN_ has specific format. Take a look at image above. You can see that _SSN_ is just only small subset of _string_ possible values.
+**Value object is precise and can validate format during object initialization.**
+
+==== Examples from different domains:
+- Iccid (International Circuit Card Identifier) in the telecommunications domain (this not just string but every character metters)
+- SSN (Social Security Number) in the financial domain
```

Comment:
Repeated + SSN is usually not a part of the financial domain ;)

Reply:
Yeah it is repetated. I don't agree with second part SSN is a critical piece of information in many financial domains in the United States. It is frequently used for e.g in credit applications.